README.md
### 整体结构

为了方便只展示 src 目录下的文件

```
├─assets            // 静态资源文件存放图片和字体文件
│  ├─font				// 字体文件 https://www.lcddjm.com/font 提取所需要的字
│  ├─image				// 图片 https://tinypng.com 压缩
|  ...
├─component       	// 每个页面所需要的组件
│  ├─Base*** 			// 基础组件
│  ├─Index*** 			// 首页所需的组件
│  └─Section*** 		// 关卡页面所需的组件     
|  ...
├─interface         // ts 所用的接口
├─pages             // 页面文件
│  ├─Choose				// 选择关卡
│  ├─Index				// 首页
│  ├─RankList			// 排行榜
│  ├─Section			// 每个关卡
│  ├─Success			// 挑战成功
└─utils              // 工具函数
    ├─convertFloatToInt	// 将浮点数转换成 10 进制数
    ├─convertNumberToUppercase // 将阿拉伯数字转换成中文数字(只个位)
    ├─convertPxToVw		// 将 px 转换成 vw (只用于牌数据，其他使用库来做)
    ├─sortCard			// 打乱牌顺序
    ├─useFetch			// 将所有请求包装成 custom hooks
    ├─useInterval		// 将 setInterval 包装成 custom hooks
    └─useTimeout		// 将 setTimeout 包装成 custom hooks
    
```

## 技术栈

- React
- react-router-dom  	// 路由
- use-react-router        // 通过 hooks 的方式获取到 react-router 的 context
- react-spring 		       // 动画
- styled-components   // CSS-in-JS
- styled-normalize       // 重置默认 CSS 样式
- styled-px2vw 			 // px to vw
- react-id-swiper           // 让 swiper 更适合 react
- swiper                          // 触摸内容滑动 js 插件



## 项目运行

使用 [Create React App](https://github.com/facebook/create-react-app) 创建，按照标准即可完成

环境变量文件 根目录 `.env`

```
REACT_APP_BE_URL=https://***
REACT_APP_GA=UA-****-*
REACT_APP_NAME=**
REACT_APP_DESC=****
REACT_APP_ICO=https://***/favicon.ico
REACT_APP_FE_URL=https://***/#/
```

## 注意点

1. 移动端 click 300 ms 延迟

   因为本项目对于点击响应速度有要求，但是使用 touch** 的方式会导致点击穿透，故使用

   ```html
   <meta name="viewport" content="user-scalable=no">
   <meta name="viewport" content="initial-scale=1,maximum-scale=1">
   ```

   禁止放大缩小

2. 点击会有高亮

   覆盖 webkit/safari, Blink/Chrome 几乎所有的移动浏览器

   ```css
   -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
   ```



## 特别的一些地方

- 使用 hooks 拿到 router 的 context

  ```js
    const {
      match: { params },
    } = useRouter<ISectionRouterProp>();
    const { step } = params;
    const [stepNum] = useState(Number(step));
  ```
  
- 使用自定义 hooks 处理数据

  - initialData 从后端拿到的数据的大体格式，这样就不需要自己写 interface 了，还解决了，没有拿到数据然后渲染出错的问题
  - postData 因为本项目只有 get 和 post 所以直接以是否有该来进行判断

  ```js
  // src/utils/useFetch.ts
  function useFetch<T>(initialUrl: string, initialData: T, postData?: any) {
    const [data, setData] = useState(initialData);
    const [url, setUrl] = useState(initialUrl);
    const [isLoading, setIsLoading] = useState(false);
    const [isError, setIsError] = useState(false);
    const didCancel = useRef<boolean>();
    useEffect(() => {
      didCancel.current = false;
      const fetchData = async () => {
        setIsError(false);
        setIsLoading(true);
        try {
          let body;
          if (postData) {
            const formBody = [];
            for (const property in postData) {
              const encodedKey = encodeURIComponent(property);
              const encodedValue = encodeURIComponent(postData[property]);
              formBody.push(`${encodedKey}=${encodedValue}`);
            }
            body = formBody.join('&');
          }
          const result = await fetch(`${process.env.REACT_APP_BE_URL}${url}`, {
            method: postData ? 'POST' : 'GET',
            body: postData ? body : null,
            headers: {
              'content-type': 'application/x-www-form-urlencoded',
              token: localStorage.getItem('id_token') || '',
            },
          });
  
          if (!didCancel.current) {
            result.json().then(a => setData(a));
          }
        } catch (error) {
          if (!didCancel.current) {
            setIsError(true);
          }
        }
  
        setIsLoading(false);
      };
      fetchData();
  
      return () => {
        didCancel.current = true;
      };
    }, [url]);
  
    const doFetch = (_url: string) => {
      setUrl(_url);
    };
  
    return {
      data,
      isLoading,
      isError,
      doFetch,
    };
  }
  ```


- 翻牌逻辑

  - 将牌的大小数据以及两牌之间的距离记录到一个数组中，以方便修改以及调用

    ```js
    // src/data/card.ts
    const cardsSize = convertPxToVw([
      {
        width: 282,
        height: 359,
        marginRight: 17,
        marginBottom: 18,
        imageHeight: 232,
        imageWidth: 245,
        fontSize: 24,
      },
      ...
    ]);
    
    ```

  ```js
  // src/component/SectionCard.tsx
  ...
    // 初始化牌的数据 
    const [cardData, setCardData] = useState(
      sortCard(cardDatas[stepNum - 1]).map((item, index) => ({
        ...item,
        isShow: false, // 是否翻开
        isHide: false, // 是否已经成功比对消失了
        id: index,
      })),
    );
    // 如上将牌的大小信息拿到
    const [cardSize] = useState(cardsSize[stepNum - 1]);
    const handelCardClick = (index: number) => {
      if (cardData[index].isHide) return; // 如果隐藏了就不执行了
      // 先查询上一次的 -> 是否已经有两张已经翻开了
      const temp = cardData.filter(item => item.isShow && !item.isHide);
      // 如果有两张已经翻开 不能新翻开牌
      if (temp.length !== 2 || cardData[index].isShow) {
        // 将牌的状态调转
        cardData[index].isShow = !cardData[index].isShow;
        // 设定新的 state
        setCardData([...cardData]);
        // 查找调整过后翻开的的牌
        const isShowCard = cardData.filter(item => item.isShow && !item.isHide);
        // 如果当前是新翻开的牌的情况 并且牌的数量为 2 检查是否满足消灭的条件
        if (cardData[index].isShow && isShowCard.length === 2) {
          if (isShowCard[0].content === isShowCard[1].content) {
            // 将牌的状态调转
            cardData[isShowCard[0].id].isHide = true;
            cardData[isShowCard[1].id].isHide = true;
            // 设定新的 state
            setCardData([...cardData]);
            // 在这种情况下查看是否还有牌没有隐藏
            // 如果没有那么就触发时间暂停
            if (!cardData.some(item => !item.isHide)) {
              const cardFinished = new Event('cardFinished');
              window.dispatchEvent(cardFinished);
              setShowMask(true);
            }
          } else {
            // 如果不满足条件，自己翻回去
            setTimeoutRef.current = setTimeout(() => {
              cardData[isShowCard[0].id].isShow = false;
              cardData[isShowCard[1].id].isShow = false;
              setCardData([...cardData]);
            }, 500);
          }
        }
      }
    };
  ...
  ```

- 组件间传递信息

  因为比较简单没有引入 redux 直接使用原生的自定义事件，并将数据存储在 localstorage 中
  
- 成功页面的彩带飞落

  参考 https://github.com/cahilfoley/react-snowfall 做了些修改


  - 使用 http://demo.qunee.com/svg2canvas/ (canvg.js) 将 svg 转换成 bezier Curve 

    ```js
    // src/snowfall/hooks.ts
       	ctx.moveTo(-0.012, 15.45);
        ctx.bezierCurveTo(-0.012, 15.45, -0.196, 25.764, 7.897, 24.97);
        ctx.bezierCurveTo(7.3, 25.054, 13.975, 9.882, 25.979, 3.028);
        ctx.bezierCurveTo(24.911, 2.685, 20.704, 5.048, 20.859, 0.802);
        ctx.bezierCurveTo(21.013, -3.443, 4.371, 9.818, -0.012, 15.45);
        ctx.closePath();
    ```

  - 然后使用 drawImage 将生成的 canvas 放到主要的 canvas 里

    ```js
    ctx.drawImage(this.image, this.params.x, this.params.y);
    ```


  ## 更新日志

  本项目遵从 [Angular Style Commit Message Conventions](https://gist.github.com/stephenparish/9941e89d80e2bc58a153)，更新日志由 `standard-changelog` 自动生成。完整日志请点击 [CHANGELOG.md](./CHANGELOG.md)

## 图例

![1569659314915](./README.assets/1569659314915.png)

![1569659472763](./README.assets/1569659472763.png)

![1569659483498](./README.assets/1569659483498.png)

![1569659514188](./README.assets/1569659514188.png)

![1569659525060](./README.assets/1569659525060.png)

![1569659533629](./README.assets/1569659533629.png)

![1570190594142](./README.assets/1570190594142.png)

CHANGELOG.md
## 0.1.1 (2019-10-04)


### Bug Fixes

* **all:** ios 点击隐藏高亮 ([9c71c5b](https://github.com/skywalker512/turn-over-great-deeds/commit/9c71c5b))
* **all:** 下一关也能点击 ([741aa87](https://github.com/skywalker512/turn-over-great-deeds/commit/741aa87))
* **all:** 修改文案 ([eb923db](https://github.com/skywalker512/turn-over-great-deeds/commit/eb923db))
* **all:** 最后一关不能返回首页 ([0a12b0c](https://github.com/skywalker512/turn-over-great-deeds/commit/0a12b0c))
* **all:** 图片修改 ([8184e56](https://github.com/skywalker512/turn-over-great-deeds/commit/8184e56))
* **all:** 增添丢失的字体 ([034a512](https://github.com/skywalker512/turn-over-great-deeds/commit/034a512))
* **all:** 字体完成 ([76c6dfb](https://github.com/skywalker512/turn-over-great-deeds/commit/76c6dfb))
* **all:** 字体缺少的字 ([49bfbf1](https://github.com/skywalker512/turn-over-great-deeds/commit/49bfbf1))
* **all:** 居中显示 ([98cbb9e](https://github.com/skywalker512/turn-over-great-deeds/commit/98cbb9e))
* **all:** 排行榜实时请求 ([c3df56e](https://github.com/skywalker512/turn-over-great-deeds/commit/c3df56e))
* **all:** 文案修改 ([a4cdc19](https://github.com/skywalker512/turn-over-great-deeds/commit/a4cdc19))
* **all:** 文案修改 ([56d93b9](https://github.com/skywalker512/turn-over-great-deeds/commit/56d93b9))
* **all:** 更新视觉图 ([b1d61ee](https://github.com/skywalker512/turn-over-great-deeds/commit/b1d61ee))
* **all:** 点击高亮全局禁止 ([1d0c42e](https://github.com/skywalker512/turn-over-great-deeds/commit/1d0c42e))
* **all:** 视觉修改 ([a6c6dbd](https://github.com/skywalker512/turn-over-great-deeds/commit/a6c6dbd))
* **all:** 视觉修改 ([9867585](https://github.com/skywalker512/turn-over-great-deeds/commit/9867585))
* **all:** 细节修改 ([3ef49da](https://github.com/skywalker512/turn-over-great-deeds/commit/3ef49da))
* **all:** 细节修改 ([5a2bf85](https://github.com/skywalker512/turn-over-great-deeds/commit/5a2bf85))
* **all:** 细节调整 ([e67f9e1](https://github.com/skywalker512/turn-over-great-deeds/commit/e67f9e1))
* **all:** 细节调整 ([2165741](https://github.com/skywalker512/turn-over-great-deeds/commit/2165741))
* **all:** 缺少的文字 ([499c2bb](https://github.com/skywalker512/turn-over-great-deeds/commit/499c2bb))
* **all:** 解决图片没居中的问题 ([384c426](https://github.com/skywalker512/turn-over-great-deeds/commit/384c426))
* **base:** 解决在渲染之前就去暂停播放 ([61a7d1d](https://github.com/skywalker512/turn-over-great-deeds/commit/61a7d1d))
* **card:** 文案修改 ([46ac477](https://github.com/skywalker512/turn-over-great-deeds/commit/46ac477))
* **card:** 细节修改 ([83ab365](https://github.com/skywalker512/turn-over-great-deeds/commit/83ab365))
* **choose:** 因为字体没有对齐的问题 ([76b59a4](https://github.com/skywalker512/turn-over-great-deeds/commit/76b59a4))
* **Choose:** 挑战文字对其 ([fcb2c3f](https://github.com/skywalker512/turn-over-great-deeds/commit/fcb2c3f))
* **index:** link 的地址 ([f23b345](https://github.com/skywalker512/turn-over-great-deeds/commit/f23b345))
* **index:** 再次进入音乐不会暂停 ([7966cd0](https://github.com/skywalker512/turn-over-great-deeds/commit/7966cd0))
* **index:** 压缩图片 ([d8bc84b](https://github.com/skywalker512/turn-over-great-deeds/commit/d8bc84b))
* **index:** 点击不会高亮 ([8897ca4](https://github.com/skywalker512/turn-over-great-deeds/commit/8897ca4))
* **index:** 细节调整 ([7ee0883](https://github.com/skywalker512/turn-over-great-deeds/commit/7ee0883))
* **index:** 解决图片被截断的问题 ([23b33f2](https://github.com/skywalker512/turn-over-great-deeds/commit/23b33f2))
* **index:** 解决微信上面表现不同 ([d6fda32](https://github.com/skywalker512/turn-over-great-deeds/commit/d6fda32))
* **index:** 首页卡片图片修改 ([2dd164e](https://github.com/skywalker512/turn-over-great-deeds/commit/2dd164e))
* **RankList:** 排行榜对其 ([f909130](https://github.com/skywalker512/turn-over-great-deeds/commit/f909130))
* **RankList:** 新的卡片 ([a7cf940](https://github.com/skywalker512/turn-over-great-deeds/commit/a7cf940))
* **router:** 出现导航栏的问题 ([fdaf850](https://github.com/skywalker512/turn-over-great-deeds/commit/fdaf850))
* **section:** 卡片对齐 ([1cd32a3](https://github.com/skywalker512/turn-over-great-deeds/commit/1cd32a3))
* **section:** 将 mask 的按钮贴到底部 ([252e80d](https://github.com/skywalker512/turn-over-great-deeds/commit/252e80d))
* **section:** 弹窗高度增加 ([53c005e](https://github.com/skywalker512/turn-over-great-deeds/commit/53c005e))
* **section:** 文案修改 ([78b8516](https://github.com/skywalker512/turn-over-great-deeds/commit/78b8516))
* **section:** 视觉修改 ([31cd9a7](https://github.com/skywalker512/turn-over-great-deeds/commit/31cd9a7))
* **section:** 细节调整 ([3dc8727](https://github.com/skywalker512/turn-over-great-deeds/commit/3dc8727))
* **section:** 细节调整 ([093b5ea](https://github.com/skywalker512/turn-over-great-deeds/commit/093b5ea))
* **section:** 细节调整 ([183ff0d](https://github.com/skywalker512/turn-over-great-deeds/commit/183ff0d))
* **snowfall:** 解决 eslint 报错 ([11c239d](https://github.com/skywalker512/turn-over-great-deeds/commit/11c239d))
* bug ([53c48db](https://github.com/skywalker512/turn-over-great-deeds/commit/53c48db))
* **popup:** 文字对齐 ([c05d960](https://github.com/skywalker512/turn-over-great-deeds/commit/c05d960))
* **section:** 没有设定 key 值 ([8d7eba7](https://github.com/skywalker512/turn-over-great-deeds/commit/8d7eba7))
* **section:** 细节调整 ([0e57f54](https://github.com/skywalker512/turn-over-great-deeds/commit/0e57f54))
* **success:** 再次出现导航栏的问题 ([eabbb61](https://github.com/skywalker512/turn-over-great-deeds/commit/eabbb61))
* **success:** 增加排行榜按钮 ([9398852](https://github.com/skywalker512/turn-over-great-deeds/commit/9398852))
* **success:** 成功界面视觉调整 ([1efd6d7](https://github.com/skywalker512/turn-over-great-deeds/commit/1efd6d7))
* **success:** 通关页居中显示 ([e35ea3f](https://github.com/skywalker512/turn-over-great-deeds/commit/e35ea3f))


### Features

* **all:** 字体引入 ([8c96a45](https://github.com/skywalker512/turn-over-great-deeds/commit/8c96a45))
* **all:** 完成主体搭建 ([f19c027](https://github.com/skywalker512/turn-over-great-deeds/commit/f19c027))
* **all:** 完成图片 ([3807ce8](https://github.com/skywalker512/turn-over-great-deeds/commit/3807ce8))

src/snowfall/hooks.ts
import React, {
  CSSProperties,
  MutableRefObject,
  useCallback,
  useEffect,
  useLayoutEffect,
  useMemo,
  useState,
} from 'react';
import Snowflake, { SnowflakeConfig } from './Snowflake';
import { snowfallBaseStyle } from './config';
import { getSize } from './utils';

const canvasSingleSnow = () => {
  const rgba = [
    { rgb: 'rgb(254, 155, 116)', rotate: 12, scale: 1 },
    { rgb: 'rgb(251, 130, 127)', rotate: 120, scale: 1.1 },
    { rgb: 'rgb(254, 249, 178)', rotate: 34, scale: 1.2 },
    { rgb: 'rgb(254, 155, 116)', rotate: 78, scale: 0.8 },
    { rgb: 'rgb(255, 189, 220)', rotate: 64, scale: 1 },
    { rgb: 'rgb(193, 206, 248)', rotate: 34, scale: 0.8 },
    { rgb: 'rgb(193, 206, 248)', rotate: 34, scale: 0.8 },
  ];
  return rgba.map(item => {
    const singleSnow = document.createElement('canvas');
    const ctx = singleSnow.getContext('2d') as CanvasRenderingContext2D;
    ctx.fillStyle = item.rgb;
    ctx.scale(item.scale, item.scale);
    ctx.rotate(item.rotate);
    ctx.beginPath();
    ctx.moveTo(-0.012, 15.45);
    ctx.bezierCurveTo(-0.012, 15.45, -0.196, 25.764, 7.897, 24.97);
    ctx.bezierCurveTo(7.3, 25.054, 13.975, 9.882, 25.979, 3.028);
    ctx.bezierCurveTo(24.911, 2.685, 20.704, 5.048, 20.859, 0.802);
    ctx.bezierCurveTo(21.013, -3.443, 4.371, 9.818, -0.012, 15.45);
    ctx.closePath();
    ctx.fill('evenodd');
    return singleSnow;
  });
};

/**
 * A utility function to create a collection of snowflakes
 * @param canvasRef A ref to the canvas element
 * @param amount The number of snowflakes
 * @param config The configuration for each snowflake
 */
const createSnowflakes = (
  canvasRef: React.MutableRefObject<HTMLCanvasElement | undefined>,
  amount: number,
  config: SnowflakeConfig,
) => {
  const snowflakes: Snowflake[] = [];
  const a = canvasSingleSnow();
  for (let i = 0; i < amount; i++) {
    const b = a[i % 7];
    snowflakes.push(
      new Snowflake(canvasRef.current as HTMLCanvasElement, config, b),
    );
  }

  return snowflakes;
};

/**
 * A utility hook to manage creating and updating a collection of snowflakes
 * @param canvasRef A ref to the canvas element
 * @param amount The number of snowflakes
 * @param config The configuration for each snowflake
 */
export const useSnowflakes = (
  canvasRef: React.MutableRefObject<HTMLCanvasElement | undefined>,
  amount: number,
  config: SnowflakeConfig,
) => {
  const [snowflakes, setSnowflakes] = useState<Snowflake[]>([]);

  // Handle change of amount
  useEffect(() => {
    setSnowflakes(() => {
      const sizeDifference = amount - snowflakes.length;

      if (sizeDifference > 0) {
        return [
          ...snowflakes,
          ...createSnowflakes(canvasRef, sizeDifference, config),
        ];
      }

      if (sizeDifference < 0) {
        return snowflakes.slice(0, amount);
      }

      return snowflakes;
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [amount, canvasRef, config]);

  // Handle change of config
  useEffect(() => {
    // eslint-disable-next-line no-shadow
    setSnowflakes(snowflakes =>
      snowflakes.map(snowflake => {
        // eslint-disable-next-line no-param-reassign
        snowflake.config = config;
        return snowflake;
      }),
    );
  }, [config]);

  return snowflakes;
};

/**
 * Returns the height and width of a HTML element, uses the `ResizeObserver` api if
 * available to detect changes to the
 * size. Falls back to listening for resize events on the window.
 * @param ref A ref to the HTML element to be measured
 */
export const useComponentSize = (
  ref: MutableRefObject<HTMLElement | undefined>,
) => {
  const [size, setSize] = useState(getSize(ref.current));

  const resizeHandler = useCallback(() => {
    if (ref.current) {
      setSize(getSize(ref.current));
    }
  }, [ref]);

  useLayoutEffect(() => {
    // @ts-ignore
    const { ResizeObserver } = window;

    if (!ref.current) return;
    resizeHandler();

    if (typeof ResizeObserver === 'function') {
      const resizeObserver = new ResizeObserver(resizeHandler);
      resizeObserver.observe(ref.current as HTMLElement);

      // eslint-disable-next-line consistent-return
      return () => resizeObserver.disconnect();
    }
    window.addEventListener('resize', resizeHandler);

    // eslint-disable-next-line consistent-return
    return () => window.removeEventListener('resize', resizeHandler);
  }, [ref, resizeHandler]);

  return size;
};

/**
 * Utility hook that merges any provided styles with the default styles
 * @param overrides The style prop passed into the component
 */
export const useSnowfallStyle = (overrides?: CSSProperties) =>
  useMemo(
    () => ({
      ...snowfallBaseStyle,
      ...(overrides || {}),
    }),
    [overrides],
  );
